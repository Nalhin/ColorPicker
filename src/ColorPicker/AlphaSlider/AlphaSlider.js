import React from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import ColorSelector from '../SliderHandle';
import { SELECTOR_HEIGHT, SELECTED_SLIDER } from '../constants';
import { colorArrayToRgba } from '../Utils';

const Background = styled.div`
  display: inline-block;
  height: 10px;
  border-radius: 4px;
  width: 100%;
`;

const Wrapper = styled.div`
  margin-top: 5px;
`;

const AlphaSlider = ({
  changeSelected,
  selected,
  alphaRef,
  alphaColor,
  alphaPosition,
  changeAlphaPosition,
  satLightColor,
}) => {
  const getColor = React.useCallback(
    event => {
      const slider = alphaRef.current;
      const rect = slider.getBoundingClientRect();
      const x = Math.floor(
        (event.clientX !== undefined
          ? event.clientX
          : event.touches[0].clientX) - rect.left,
      );

      if (x <= 0) changeAlphaPosition({ x: 0, y: SELECTOR_HEIGHT });
      else if (x >= slider.offsetWidth)
        changeAlphaPosition({ x: slider.offsetWidth, y: SELECTOR_HEIGHT });
      else changeAlphaPosition({ x, y: SELECTOR_HEIGHT });
    },
    [alphaRef, changeAlphaPosition],
  );

  const getPositionMove = React.useCallback(
    event => {
      event.preventDefault();
      if (selected === SELECTED_SLIDER.ALPHA) {
        getColor(event);
      }
    },
    [getColor, selected],
  );

  const getPositionClick = event => {
    event.preventDefault();
    getColor(event);
    changeSelected(SELECTED_SLIDER.ALPHA);
  };

  React.useEffect(() => {
    if (selected === SELECTED_SLIDER.ALPHA) {
      window.addEventListener('mousemove', getPositionMove);
      window.addEventListener('ontouchmove', getPositionMove);
    } else {
      window.removeEventListener('mousemove', getPositionMove);
      window.removeEventListener('ontouchmove', getPositionMove);
    }
    return () => {
      window.removeEventListener('mousemove', getPositionMove);
      window.removeEventListener('ontouchmove', getPositionMove);
    };
  }, [getPositionMove, selected]);

  return (
    <Wrapper>
      <ColorSelector
        color={alphaColor}
        position={alphaPosition}
        click={getPositionClick}
      >
        <Background
          style={{
            height: '10px',
            background: `linear-gradient(90deg,rgba(0, 0, 0, 0), ${colorArrayToRgba(
              satLightColor,
            )}) `,
          }}
          ref={alphaRef}
        />
      </ColorSelector>
    </Wrapper>
  );
};

AlphaSlider.propTypes = {
  changeSelected: PropTypes.func.isRequired,
  selected: PropTypes.number.isRequired,
  alphaRef: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.shape({ current: PropTypes.instanceOf(Element) }),
  ]).isRequired,
  alphaColor: PropTypes.arrayOf(PropTypes.number).isRequired,
  alphaPosition: PropTypes.shape({ x: PropTypes.number, y: PropTypes.number })
    .isRequired,
  changeAlphaPosition: PropTypes.func.isRequired,
  satLightColor: PropTypes.arrayOf(PropTypes.number).isRequired,
};

export default AlphaSlider;
