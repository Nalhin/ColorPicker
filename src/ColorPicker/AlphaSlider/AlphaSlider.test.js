import React from 'react';
import { render } from '@testing-library/react';

import AlphaSlider from './AlphaSlider';

describe('Alpha Slider ', () => {
  const props = {
    selected: 1,
    changeSelected: () => {},
    alphaRef: () => {},
    alphaColor: [255, 240, 220, 1],
    alphaPosition: { x: 40, y: 50 },
    changeAlphaPosition: () => {},
    satLightColor: [255, 240, 220, 1],
  };
  it('Should render correctly', () => {
    const { container } = render(<AlphaSlider {...props} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
