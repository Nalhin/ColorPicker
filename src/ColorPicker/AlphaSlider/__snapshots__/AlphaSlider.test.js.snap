// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alpha Slider  Should render correctly 1`] = `
<div
  class="css-1fh398w-AlphaSlider__Wrapper e1qunp8u1"
>
  <div
    class="css-cqd1um-SliderHandle__Wrapper e9y9zbe0"
  >
    <div
      class="css-gjeaqo-SliderHandle__Handle e9y9zbe1"
      style="top: 50px; left: 33px; background: rgb(255, 240, 220);"
    />
    <div
      class="css-2fon4n-AlphaSlider__Background e1qunp8u0"
      style="height: 10px;"
    />
  </div>
</div>
`;
